text/0000-import-trait-associated-functions.md

```diff
@@ -0,0 +1,235 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Importing an associated functions from a trait is the same as importing a function from any module:
+```rust
+use Default::default;
+```
+
+Once you have done this, the function is made available in your current scope just like any other regular function.
+
+```rust
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+
+You can also use this with trait methods (i.e. associated functions that take a `self` argument):
+
+```rust
+use num_traits::float::Float::{sin, cos}
+
+fn eulers_formula(theta: f64) -> (f64, f64) {
+    (cos(theta), sin(theta))
+}
+```
+
+Importing an associated function from a trait does not import the trait. If you want to call `self` methods on a trait or `impl` it, then you can import the trait and its associated functions in a single import statement:
+
+```rust
+mod a {
+    trait A {
+        fn new() -> Self;
+        fn do_something(&self);
+    }
+}
+
+mod b {
+    use super::a::A::{self, new}
+
+    struct B();
+
+    impl A for B {
+        fn new() -> Self {
+            B()
+        }
+
+        fn do_something(&self) {
+        }
+    }
+
+    fn f() {
+        let b: B = new();
+        b.do_something();
+    }
+}
+```
+
+Associated functions can also be renamed when they are imported using the usual `as` syntax:
+```rust
+use Default::default as gimme
+
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: gimme()
+        }
+    }
+}
+```
+
+You cannot import a parent trait associated function from a sub-trait:
+
+```rust
+use num_traits::float::Float::zero; // Error: try `use num_traits::identities::Zero::zero` instead.
+
+fn main() {
+    let x : f64 = zero();
+    println!("{}",x);
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When 
+
+```rust
+use Trait::func as m;
+```
+occurs, a new item `m` is made available in the value namespace of the current module. Any attempts to call this item are treated calling the associated function explicitly qualified. As always, the `as` qualifier is optional, in which case the name of the new item is identical with the name of the associated function in the trait. In other words, the example:
+
+```rust
+use Default::default;
+
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+desugars to 
+```rust
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: Default::default()
+        }
+    }
+}
+```
+And a call
+```rust
+use Trait::func as m;
+m(x, y, z);
+```
+desugars to
+```rust
+Trait::func(x, y, z);
+```
+
+Additionally, the syntax
+```rust
+use some_module::Trait::self;
```

Comment:
Perhaps jump straight to a `use some_module::Trait::{self, method};` example, since that's the goal of this syntax?

Reply:
Actually, on second though, that is already allowed
```rs
trait T {}

mod module {
    use crate::T::{self};
}
```
compiles today as expected. So yes, it would make more sense to just jump straight to the `use some_module::Trait::{self, method};` example and drop the rest.